guidance/models/transformers/_transformers.py

```diff
@@ -132,13 +144,50 @@ def _tokenizer(self, model, **kwargs):
 
         return tokenizer
 
-    def __call__(self, byte_string):
+    def encode(self, byte_string: bytes) -> Sequence[int]:
+        assert isinstance(byte_string, bytes)
         tokenisation = self._orig_tokenizer(byte_string)
         return tokenisation["input_ids"]
 
+    def decode(self, tokens: Sequence[int]) -> bytes:
+        decoded_str = self._orig_tokenizer.decode(tokens)
+        return decoded_str.encode()
+    
+    def recode( self, tokens: Sequence[int]) -> Sequence[int]:
+        # the encode/decode cycle might not work if we have partial unicode strings
+        used_tokens = len(tokens)
+        for _ in range(3):
```

Comment:
Why range(3)? Can you document what's going on here a bit more?

Reply:
I know this is just a refactor and not new code, but additional comments would still be nice :)

Reply:
I will see if I can figure that out....

Reply:
If you don't know, feel free to punt on the issue...

Reply:
Between this part and the next little block, it would appear that a unicode character might be split between two tokens @Harsha-Nori ? In that case, the decode to UTF-8 would fail, since the final character would be incomplete (and I might be _really_ mixing up some terminology here). That then implies to me that there are some tokens which can only be followed by a subset of the other tokens (since there would only be certain valid completions of the split token).

I'm not sure why it limits the step-back to three, though.

Reply:
Well, some [further research](https://en.wikipedia.org/wiki/UTF-8) says that the maximum size of a ~character~ codepoint in UTF-8 is 4 bytes (according to [another reference](https://www.joelonsoftware.com/2003/10/08/the-absolute-minimum-every-software-developer-absolutely-positively-must-know-about-unicode-and-character-sets-no-excuses/), six bytes are theoretically possible, but not currently supported).

That would tend to suggest that we might have to back up _four_ tokens (if we assume that the set of tokens included single-byte tokens), so not quite sure why it's three (although I'm also suspecting that if you have to delete an entire ~character~ codepoint from the end of the string, then you're looking at something the model could _never_ have generated).